plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/version/KubernetesVersionManagerImpl.java

```diff
@@ -0,0 +1,377 @@
+
+        return createKubernetesSupportedVersionListResponse(versions);
+    }
+
+    @Override
+    @ActionEvent(eventType = KubernetesVersionEventTypes.EVENT_KUBERNETES_VERSION_ADD, eventDescription = "Adding Kubernetes supported version")
+    public KubernetesSupportedVersionResponse addKubernetesSupportedVersion(final AddKubernetesSupportedVersionCmd cmd) {
+        if (!KubernetesClusterService.KubernetesServiceEnabled.value()) {
+            throw new CloudRuntimeException("Kubernetes Service plugin is disabled");
+        }
+        String name = cmd.getName();
+        final String semanticVersion = cmd.getSemanticVersion();
+        final Long zoneId = cmd.getZoneId();
+        final Long isoId = cmd.getIsoId();
+        final String isoUrl = cmd.getUrl();
+        final String isoChecksum = cmd.getChecksum();
+        if (compareSemanticVersions(semanticVersion, MIN_KUBERNETES_VERSION) < 0) {
+            throw new InvalidParameterValueException(String.format("New supported Kubernetes version cannot be added as %s is minimum version supported by Kubernetes Service", MIN_KUBERNETES_VERSION));
+        }
+        if (Strings.isNullOrEmpty(isoUrl) && (isoId == null || isoId <= 0)) {
+            throw new InvalidParameterValueException(String.format("Either %s or %s parameter must be passed to add a new supported Kubernetes version", "isourl", ApiConstants.ISO_ID));
+        }
+        if (!Strings.isNullOrEmpty(isoUrl) && isoId != null && isoId > 0) {
+            throw new InvalidParameterValueException(String.format("Both %s and %s parameters can not be passed simultaneously to add a new supported Kubernetes version", ApiConstants.URL, ApiConstants.ISO_ID));
+        }
+        if (zoneId != null && dataCenterDao.findById(zoneId) == null) {
+            throw new InvalidParameterValueException("Invalid zone specified");
+        }
+        if (Strings.isNullOrEmpty(name)) {
+            name = String.format("v%s", semanticVersion);
+            if (zoneId != null) {
+                name = String.format("%s-%s", name, dataCenterDao.findById(zoneId).getName());
+            }
+        }
+
+        VMTemplateVO template = null;
+        if (isoId != null) {
+            template = templateDao.findById(isoId);
+        }
+        if (template == null) { // register new ISO
+            VirtualMachineTemplate vmTemplate = null;
+            try {
+                vmTemplate = registerKubernetesVersionIso(name, isoUrl, isoChecksum);
+            } catch (IllegalAccessException | NoSuchFieldException | IllegalArgumentException | ResourceAllocationException ex) {
+                LOGGER.error(String.format("Unable to register binaries ISO for supported kubernetes version, %s", name), ex);
+                throw new CloudRuntimeException(String.format("Unable to register binaries ISO for supported kubernetes version, %s", name));
+            }
+            template = templateDao.findById(vmTemplate.getId());
```

Comment:
this comment basically says: thios is a method that should be called registerNewIso() and factorred out.  the name can be improved though. a bit larger scope can be to include the isoId != null check and call the method registerKubernetesISOIfNeeded().

Reply:
refactored